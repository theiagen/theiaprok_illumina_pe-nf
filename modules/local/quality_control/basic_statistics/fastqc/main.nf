process FASTQC {
    tag "$meta.id"
    label "process_low"

    container "us-docker.pkg.dev/general-theiagen/staphb/fastqc:0.12.1"

    input:
    tuple val(meta), path(read1), path(read2)

    output:
    tuple val(meta), path("*_R1_fastqc.html"), emit: read1_fastqc_html
    tuple val(meta), path("*_R1_fastqc.zip"), emit: read1_fastqc_zip
    tuple val(meta), path("*_R2_fastqc.html"), emit: read2_fastqc_html
    tuple val(meta), path("*_R2_fastqc.zip"), emit: read2_fastqc_zip
    tuple val(meta), path("READ1_SEQS.txt"), emit: read1_seq_txt
    tuple val(meta), path("READ2_SEQS.txt"), emit: read2_seq_txt  
    tuple val(meta), path("READ_PAIRS.txt"), emit: read_pairs_txt
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    
    // Extract base names for reads
    def read1_name = read1.name.replaceAll(/\.gz$/, '').replaceAll(/\.(fastq|fq)$/, '')
    def read2_name = read2.name.replaceAll(/\.gz$/, '').replaceAll(/\.(fastq|fq)$/, '')
    
    """
    # run fastqc: 
    # --extract: uncompress output files
    fastqc \\
        --outdir . \\
        --threads ${task.cpus} \\
        --extract \\
        ${read1} \\
        ${read2} \\
        ${args}
    
    # Rename output files to include prefix
    mv ${read1_name}_fastqc.html ${prefix}_R1_fastqc.html
    mv ${read1_name}_fastqc.zip ${prefix}_R1_fastqc.zip
    mv ${read2_name}_fastqc.html ${prefix}_R2_fastqc.html
    mv ${read2_name}_fastqc.zip ${prefix}_R2_fastqc.zip
    
    # Extract sequence counts
    read1_seqs=\$(grep "Total Sequences" ${read1_name}_fastqc/fastqc_data.txt | cut -f 2)
    read2_seqs=\$(grep "Total Sequences" ${read2_name}_fastqc/fastqc_data.txt | cut -f 2)
    
    # capture number of read pairs
    if [ "\${read1_seqs}" == "\${read2_seqs}" ]; then
        read_pairs="\${read1_seqs}"
    else
        read_pairs="Uneven pairs: R1=\${read1_seqs}, R2=\${read2_seqs}"
    fi
    
    # Write values to files for outputs
    echo "\${read1_seqs}" > READ1_SEQS.txt
    echo "\${read2_seqs}" > READ2_SEQS.txt
    echo "\${read_pairs}" > READ_PAIRS.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastqc: \$(fastqc --version | sed 's/FastQC v//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_R1_fastqc.html
    touch ${prefix}_R1_fastqc.zip
    touch ${prefix}_R2_fastqc.html
    touch ${prefix}_R2_fastqc.zip
    
    echo "1000" > READ1_SEQS.txt
    echo "1000" > READ2_SEQS.txt
    echo "1000" > READ_PAIRS.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastqc: 0.12.1
    END_VERSIONS
    """
}