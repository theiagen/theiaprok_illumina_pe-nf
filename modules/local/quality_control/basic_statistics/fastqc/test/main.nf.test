nextflow_process {
    name "Test Process FASTQC"
    script "../main.nf"
    process "FASTQC"
    
    test("fastqc test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                {  assert snapshot(
                    process.out.read1_fastqc_html,
                    process.out.read1_fastqc_zip,
                    process.out.read2_fastqc_html,
                    process.out.read2_fastqc_zip,
                    process.out.read1_seq_txt,
                    process.out.read2_seq_txt,
                    process.out.read_pairs_txt,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("fastqc - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test_sample' ],
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_1.fastq.gz", checkIfExists: true),
                    file(params.pipelines_testdata_base_path + "genomics/prokaryotes/bacteroides_fragilis/illumina/fastq/test1_2.fastq.gz", checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}