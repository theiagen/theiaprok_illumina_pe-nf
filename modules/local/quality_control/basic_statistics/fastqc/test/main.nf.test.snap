{
    "fastqc test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R1_fastqc.html:md5,486ad4a0df6218d0d68b52e174671f74"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R1_fastqc.zip:md5,880f7d557082e6fdda2f92a8cdf0a376"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R2_fastqc.html:md5,282db1d244d50c7bf9b491f9a45cc84a"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample_R2_fastqc.zip:md5,f1483f86e66ef223a7ea67d0167ccd07"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "READ1_SEQS.txt:md5,2ebd3caa45b31a2e74e436b645eab4b0"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "READ2_SEQS.txt:md5,2ebd3caa45b31a2e74e436b645eab4b0"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "READ_PAIRS.txt:md5,2ebd3caa45b31a2e74e436b645eab4b0"
                ]
            ],
            [
                "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-06-17T16:30:45.448878"
    },
    "fastqc - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R1_fastqc.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R1_fastqc.zip:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R2_fastqc.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R2_fastqc.zip:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ1_SEQS.txt:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ2_SEQS.txt:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ_PAIRS.txt:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "7": [
                    "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
                ],
                "read1_fastqc_html": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R1_fastqc.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "read1_fastqc_zip": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R1_fastqc.zip:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "read1_seq_txt": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ1_SEQS.txt:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "read2_fastqc_html": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R2_fastqc.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "read2_fastqc_zip": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_R2_fastqc.zip:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "read2_seq_txt": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ2_SEQS.txt:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "read_pairs_txt": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "READ_PAIRS.txt:md5,ad865d2f63b9feb2552c220385fbb7e3"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,e1cc25ca8af856014824abd842e93978"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-06-17T16:30:57.587586"
    }
}